src/Common/src/Common.Expression/Spring/Ast/SpelNode.cs

```diff
@@ -140,95 +139,191 @@ protected internal virtual IValueRef GetValueRef(ExpressionState state)
+                }
+            }
+            else
+            {
+                for (var i = 0; i < paramDescriptors.Length; i++)
+                {
+                    GenerateCodeForArgument(gen, cf, arguments[i], paramDescriptors[i]);
+                }
+            }
+        }
+
+        protected static void GenerateCodeForArgument(ILGenerator gen, CodeFlow cf, SpelNode argument, TypeDescriptor paramDesc)
+        {
+            cf.EnterCompilationScope();
+            argument.GenerateCode(gen, cf);
+            var lastDesc = cf.LastDescriptor();
+            if (lastDesc == null)
+            {
+                throw new InvalidOperationException("No last descriptor");
+            }
+
+            var valueTypeOnStack = CodeFlow.IsValueType(lastDesc);
+
+            // Check if need to box it for the method reference?
+            if (valueTypeOnStack && paramDesc.IsReferenceType)
+            {
+                CodeFlow.InsertBoxIfNecessary(gen, lastDesc);
+            }
+            else if (paramDesc.IsValueType && !paramDesc.IsBoxed && !valueTypeOnStack)
+            {
+                gen.Emit(OpCodes.Unbox_Any, paramDesc.Value);
+            }
+            else
+            {
+                // This would be unnecessary in the case of subtyping (e.g. method takes Number but Integer passed in)
+                CodeFlow.InsertCastClass(gen, paramDesc);
+            }
+
+            cf.ExitCompilationScope();
+        }
+
+        protected static OpCode GetLdElemInsn(Type arrElemType)
+        {
+            if (arrElemType == typeof(sbyte))
+            {
+                return OpCodes.Ldelem_I1;
+            }
+
+            if (arrElemType == typeof(byte))
+            {
+                return OpCodes.Ldelem_U1;
+            }
+
+            if (arrElemType == typeof(short) || arrElemType == typeof(char))
+            {
+                return OpCodes.Ldelem_I2;
+            }
+
+            if (arrElemType == typeof(ushort))
+            {
+                return OpCodes.Ldelem_U2;
+            }
+
+            if (arrElemType == typeof(int))
+            {
+                return OpCodes.Ldelem_I4;
+            }
+
+            if (arrElemType == typeof(uint))
+            {
+                return OpCodes.Ldelem_U4;
+            }
+
+            if (arrElemType == typeof(long))
+            {
+                return OpCodes.Ldelem_I8;
+            }
+
+            if (arrElemType == typeof(ulong))
+            {
+                return OpCodes.Ldelem_I8;
+            }
+
+            if (arrElemType == typeof(float))
+            {
+                return OpCodes.Ldelem_R4;
+            }
+
+            if (arrElemType == typeof(double))
+            {
+                return OpCodes.Ldelem_R8;
+            }
+
+            if (arrElemType == typeof(IntPtr) || arrElemType == typeof(UIntPtr))
+            {
+                return OpCodes.Ldelem_I;
+            }
+
+            return OpCodes.Ldelem_Ref;
+        }
```

Comment:
```suggestion
        protected static OpCode GetLdElemInsn(object arrElemType)
        {
            return arrElemType switch
            {
                sbyte => OpCodes.Ldelem_I1,
                byte => OpCodes.Ldelem_U1,
                short or char => OpCodes.Ldelem_I2,
                ushort => OpCodes.Ldelem_U2,
                int => OpCodes.Ldelem_I4,
                uint => OpCodes.Ldelem_U4,
                long => OpCodes.Ldelem_I8,
                ulong => OpCodes.Ldelem_I8,
                float => OpCodes.Ldelem_R4,
                double => OpCodes.Ldelem_R8,
                IntPtr or UIntPtr => OpCodes.Ldelem_I,
                _ => OpCodes.Ldelem_Ref,
            };
        }
```
This is what it might look like, but I'm seeing a test failure in `Steeltoe.Common.Expression.Internal.Spring.SpelCompilationCoverageTests.Indexer` with this change... I had to change the param to `object` to get it to compile and it's always hitting the default (as a result?) so something is wrong with this version

Reply:
Found a workaround for Type not being supported in `switch` here: https://github.com/dotnet/csharplang/discussions/356#discussioncomment-457722

Reply:
Pushed changes